tools/ci/flink-ci-tools/src/main/java/org/apache/flink/tools/ci/optional/ShadeOptionalChecker.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.flink.tools.ci.utils.shared.DependencyTree;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Verifies that all dependencies bundled with the shade-plugin are marked as optional in the pom.
+ * This ensures compatibility with later maven versions and in general simplifies dependency
+ * management as transitivity is no longer dependent on the shade-plugin.
+ */
+public class ShadeOptionalChecker {
+    private static final Logger LOG = LoggerFactory.getLogger(ShadeOptionalChecker.class);
+
+    public static void main(String[] args) throws IOException {
+        if (args.length < 2) {
+            System.out.println(
+                    "Usage: ShadeOptionalChecker <pathShadeBuildOutput> <pathMavenDependencyOutput>");
+            System.exit(1);
+        }
+
+        final Path shadeOutputPath = Paths.get(args[0]);
+        final Path dependencyOutputPath = Paths.get(args[1]);
+
+        final Map<String, Set<Dependency>> bundledDependenciesByModule =
+                ShadeParser.parseShadeOutput(shadeOutputPath);
+        final Map<String, DependencyTree> dependenciesByModule =
+                DependencyParser.parseDependencyTreeOutput(dependencyOutputPath);
+
+        final Multimap<String, Dependency> violations = ArrayListMultimap.create();
+
+        for (String module : bundledDependenciesByModule.keySet()) {
+            LOG.debug("Checking module '{}'.", module);
+            if (!dependenciesByModule.containsKey(module)) {
+                throw new IllegalStateException(
+                        String.format(
+                                "Module %s listed by shade-plugin, but not dependency-plugin.",
+                                module));
+            }
+
+            Collection<Dependency> bundledDependencies =
+                    bundledDependenciesByModule.get(module).stream()
+                            // force-shading isn't relevant for this check but breaks some shortcuts
+                            .filter(
+                                    dependency ->
+                                            !dependency
+                                                    .getArtifactId()
+                                                    .equals("flink-shaded-force-shading"))
+                            .collect(Collectors.toSet());
+
+            if (bundledDependencies.isEmpty()) {
+                LOG.debug("\tModule is not bundling any dependencies.");
+                continue;
+            }
+
+            final DependencyTree dependencyTree = dependenciesByModule.get(module);
+
+            final List<Dependency> directTransitiveDependencies =
```

Comment:
This early exit check doesn't really make sense. If something is bundled then it should have some direct transitive dependency. Otherwise there's usually nothing to bundle (e.g., when everything is set to provided).
This probably just introduces inconsistencies in the optional flags for some module without providing any real value.